druid/src/app_delegate.rs

```diff
@@ -57,46 +55,18 @@ type EventFn<T> = dyn Fn(Event, &mut T, &Env, &mut DelegateCtx) -> Option<Event>
 /// It is a natural place for things like window and menu management.
 ///
 /// You customize the `AppDelegate` by passing closures during creation.
-pub struct AppDelegate<T> {
-    event_fn: Option<Box<EventFn<T>>>,
-}
-
-impl<T: Data> AppDelegate<T> {
-    /// Create a new `AppDelegate`.
-    pub fn new() -> Self {
-        AppDelegate { event_fn: None }
-    }
-
-    /// Set the `AppDelegate`'s event handler. This function receives all events,
+pub trait AppDelegate<T: Data> {
+    /// The `AppDelegate`'s event handler. This function receives all events,
     /// before they are passed down the tree.
     ///
     /// The return value of this function will be passed down the tree. This can
     /// be the even that was passed in, a different event, or no event. In all cases,
     /// the `update` method will be called as usual.
-    pub fn event_handler<F>(mut self, f: F) -> Self
-    where
-        F: Fn(Event, &mut T, &Env, &mut DelegateCtx) -> Option<Event> + 'static,
-    {
-        self.event_fn = Some(Box::new(f));
-        self
-    }
-
-    pub(crate) fn event(
+    fn event(
         &mut self,
         event: Event,
         data: &mut T,
         env: &Env,
         ctx: &mut DelegateCtx,
-    ) -> Option<Event> {
-        match self.event_fn.as_ref() {
-            Some(f) => (f)(event, data, env, ctx),
-            None => Some(event),
-        }
-    }
-}
-
-impl<T: Data> Default for AppDelegate<T> {
-    fn default() -> Self {
-        AppDelegate::new()
-    }
+    ) -> Option<Event>;
```

Comment:
I would give this a default impl, which which is just `Some(event)`, e.g. it just passes through the event unchanged.

My thinking is that, on the app delegate, all methods are optional.

Reply:
Ah, yes, I had this thought too, but I've just iterated through several redesigns and totally forgot it. I'll add that one for sure. :+1:

Reply:
So the compiler complains about unused variables, but I don't want to underscore the names, as they'll be part of the docs. Ergo the current approach, as suggested in https://github.com/rust-lang/rust/issues/26487.